alphazero/_game.py

```diff
@@ -383,3 +384,339 @@ def n_players():
+
+        return new_board
+
+    @staticmethod
+    def get_initial_board(height=6, width=7):
+        """
+        Get the initial game board for this game. For Connect 4, for instance,
+        this should be an empty board of width `width` and height `height`, but
+        for chess this may be the initial position of the pieces.
+        """
+        return torch.full((height, width), -1)
+
+    @staticmethod
+    def get_legal_moves(board) -> torch.Tensor:
+        """
+        Get a list of legal moves for the given board game position. Should
+        return a `torch.Tensor` full of booleans that represent whether a move
+        is valid or not. Does not affect internal state of any instances of
+        this class.
+        """
+        return board[0, :] == -1
+
+    def current_legal_moves(self) -> torch.Tensor:
+        """
+        Get a dynamic list of legal moves for the current game position. Should
+        return a `torch.Tensor` full of booleans that represent whether a move
+        is valid or not.
+        """
+        return self.get_legal_moves(self._board)
+
+    @staticmethod
+    def winning_move(board, col) -> bool:
+        """
+        Check if the last move on columns `col` lead to victory
+        """
+        height = board.shape[0]
+        width = board.shape[1]
+
+        row = height - 1
+        while row >= 0 and board[row, col] != -1:
+            row -= 1
+        row += 1
+        player = board[row, col]
+        if height - row >= 4:
+            # check vertically
+            r = row - 1
+            while r > height and board[r, col] == player:
+                r -= 1
+            if r - row >= 4:
+                return True
+
+        # check horizontally
+        c1 = col - 1
+        while c1 >= 0 and board[row, c1] == player:
+            c1 -= 1
+        c2 = col + 1
+        while c2 < width and board[row, c2] == player:
+            c2 += 1
+        if c2 - c1 > 4:
+            return True
+
+        # check first diagonal
+        d1 = 1
+        while (
+            row - d1 >= 0
+            and col - d1 >= 0
+            and board[row - d1, col - d1] == player
+        ):
+            d1 += 1
+        d2 = 1
+        while (
+            row + d2 < height
+            and col + d2 < width
+            and board[row + d2, col + d2] == player
+        ):
+            d2 += 1
+        if d2 + d1 > 4:
+            return True
+
+        # check second diagonal
+        d1 = 1
+        while (
+            row - d1 >= 0
+            and col + d1 < width
+            and board[row - d1, col + d1] == player
+        ):
+            d1 += 1
+        d2 = 1
+        while (
+            row + d2 < height
+            and col - d2 >= 0
+            and board[row + d2, col - d2] == player
+        ):
+            d2 += 1
+        if d2 + d1 > 4:
+            return True
+
+        return False
+
+    @staticmethod
+    def get_game_status(board) -> tuple:
```

Comment:
same as winning_move

Reply:
Tested as much as I could, it works for all rectangular boards and I got rid of a few extraneous variables that I was creating(It should behave O(n^2) now, it might have been closer to cubic earlier sorry).